### PR TITLE
[mlir][scf]: Avoid inserting affine.min when tiling dynamic operation sizes if possible

### PR DESCRIPTION
**Pull request by AviadCo**
* During operation tiling using scf, we may avoid inserting affine.min to handle the last tile where `upper_bound = step * k` where stride may be a constant or a dynamic.

### CODE REVIEW

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 1]
```diff
@@ -186,11 +186,35 @@ static void checkSafeToTileToForall(TilingInterface op,
   }
 }
 
+/// Returns true if `size` is dynamic multiplication of `stride`.
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:189]: Not sure I follow. The `tileDividesIterationDomain` is checking for (ub - lb) % tilesize = 0. Why doesnt that catch the case you are trying to cover here?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Would it be possible to this as a cleanup on `

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp (hunk 1 above)
Comment (AviadCo) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:189]: @MaheshRavishankar 
you are right, my change adds dynamic support for `(ub - lb) % tilesize` where `lb` is zero, and ub is dynamicly defined by `possibly_dynamic_ub = possibly_dynamic_tilesize * possibly_dynamic_stride`.
Some background for this change - I am using split on dynamic operations (i.e. `linalg::generic`) where the I use `arith.divui` to calculate a tile which fits my cache and another part of `arith.remui` if the dynamic size won't be perfectly divided.
After the split, the first part size is defined by `arith.muli` where one of its operands is the tile size I am using to tile the operation - but the the `ub` is now considered dynamic although we can know the tile size divides it.

Comment (MaheshRavishankar) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:189]: Ok, I see what you are trying. I am not a fan of this kind of piecemeal approaches. I tried something. I changed your input to use affine maps. Still didnt work. I get something like this
```
%1 = affine.apply affine_map<()[s0] -> ((s0 floordiv 80) * 80)>()[%dim]
%3 = scf.for %arg2 = %c0 to %1 step %c80 iter_args(%arg3 = %extracted_slice_1) -> (tensor<?xi16>) {
      %5 = affine.min affine_map<(d0)[s0] -> (-d0 + (s0 floordiv 80) * 80, 80)>(%arg2)[%dim]
```

IIRC adding `--scf-for-loop-canonicalization` should have fixed the problem, but it didnt

cc @matthias-springer  I thought this was fixed at some time.

Anyway, I am OK with this patch overall given where things are, but this needs to account for the lower bound of the loop as well.

Comment (matthias-springer) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:189]: I haven't looked into this pass for a long time. What I vaguely remember is that `-scf-for-loop-canonicalization` has always been a bit fragile due to limitations around semi-affine expressions in `FlatAffineValueConstraints`.

Comment (AviadCo) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp]: @MaheshRavishankar thanks for the replay!
I tried to take a look at `scf-for-loop-canonicalization` which I was unfamiliar with, as for now I didn't manage to find the issue why it does not manage to remove the `affine.min` in this case. I updated the MR to also handle none `0` offset.

Comment (AviadCo) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp]: @MaheshRavishankar can you please take a look on the latest patch?

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 2]
```diff
@@ -186,18 +188,49 @@ static void checkSafeToTileToForall(TilingInterface op,
...
-  if (!offsetAsInt)
-    return false;
   std::optional<int64_t> sizeAsInt = getConstantIntValue(loopRange.size);
-  if (!sizeAsInt)
-    return false;
-  std::optional<int64_t> strideAsInt = getConstantIntValue(loopRange.stride);
-  if (!strideAsInt)
-    return false;
-  return ((sizeAsInt.value() - offsetAsInt.value()) % strideAsInt.value() == 0);
+  if (strideAsInt && offsetAsInt && sizeAsInt)
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:211]: nit: Please add `{` `}` around multi-line statements.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 3]
```diff
@@ -186,18 +188,49 @@ static void checkSafeToTileToForall(TilingInterface op,
...
+                            SmallVector<OpFoldResult> &dividers) {
+  dividers.push_back(ofr);
+  if (ofr.is<Attribute>())
+    return;
+  auto mulOp = cast<Value>(ofr).getDefiningOp<arith::MulIOp>();
+  if (!mulOp)
+    return;
+
+  // Given `ofr` = `x` * `y`, all dividers of `x` and `y` are dividers of `ofr`.
+  collectDividers(mulOp.getLhs(), dividers);
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:202]: Recursions are generally discouraged.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 4]
```diff
@@ -186,18 +188,49 @@ static void checkSafeToTileToForall(TilingInterface op,
...
+  // `offset`.
+  auto isStrideDividesDivider = [&](OpFoldResult divider) {
+    if (!strideAsInt)
+      // `stride` is dynamic.
+      return divider == loopRange.stride;
+
+    std::optional<int64_t> dividerAsInt = getConstantIntValue(divider);
+    return dividerAsInt && *dividerAsInt % *strideAsInt == 0;
+  };
+  return llvm::any_of(dividersOfSize, isStrideDividesDivider) &&
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:232]: Dont you think this logic is more complicated that you want it to be. I'd like to scope this a bit more narrowly. Just look for the immediate operations being a `arith.mul` and avoid the recursion.

The more general solution should be using something the `IntegerDivisibilityAnalysis` ([here](https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Dialect/Util/Analysis/IntegerDivisibilityAnalysis.h)) that is very similar to the range inference analysis that can be used to fold this away. Doing this in transformations like this becomes unmaintainable in the long run.

**Review by MaheshRavishankar — APPROVED** [review]
Ok, this seems to be doing much more than what the original PR set out to do. I am fine with it, but I'd really like this logic simpler. I cant gaurantee that as written this is maintainable.

**Comment by AviadCo** [comment]
@MaheshRavishankar per the discussion with you I decided to abandon the approach of searching patterns in arith operations as it may be too fragile like you mentioned. Instead I am avoiding the "split" before tiling and handling the `affine.min` accordingly.
I will also investigate `IntegerDivisibilityAnalysis ` &`scf-for-loop-canonicalization`  as those seems intresting for my flow, I appreciate your review!

**Pull request closed without merging**